File: Algorithms/SearchingAlgorithms/BFS/py/BFS.py
from collections import deque

def bfs(graph, start):
    """
    Perform Breadth-First Search on a graph
    
    Args:
    graph: dict - adjacency list representation of the graph
    start: starting node for BFS
    
    Returns:
    tuple: (visited nodes in order, parent dictionary)
    """
    visited = []  # List to keep track of visited nodes in order
    queue = deque()  # Initialize a queue
    parent = {start: None}  # To keep track of the parent of each node
    
    queue.append(start)
    
    while queue:
        node = queue.popleft()
        
        if node not in visited:
            visited.append(node)
            
            # Explore all neighbors
            for neighbor in graph[node]:
                if neighbor not in parent:  # If not visited yet
                    parent[neighbor] = node
                    queue.append(neighbor)
    
    return visited, parent

# Example usage:
if __name__ == "__main__":
    # Sample graph represented as adjacency list
    graph = {
        'A': ['B', 'C'],
        'B': ['A', 'D', 'E'],
        'C': ['A', 'F'],
        'D': ['B'],
        'E': ['B', 'F'],
        'F': ['C', 'E']
    }
    
    start_node = 'A'
    visited_nodes, parent_dict = bfs(graph, start_node)
    
    print("BFS Visited Order:", visited_nodes)
    print("Parent Relationships:", parent_dict)
    
    # To reconstruct path from start to any node:
    def get_path(parent, node):
        path = []
        while node is not None:
            path.append(node)
            node = parent[node]
        return path[::-1]  # Reverse to get start->end order
    
    print("Path from A to F:", get_path(parent_dict, 'F'))

File: Algorithms/BFS/java/Bfs.java
package BFS.java;

import java.util.*;

public class Bfs {
    public static class BFSresult {
        List<String> visitedorder;
        Map<String, String> parentMap;

        public BFSresult(List<String> visitedorder, Map<String, String> parentMap) {
            this.visitedorder = visitedorder;
            this.parentMap = parentMap;
        }
    }

    public static BFSresult bfs(Map<String, List<String>> graph, String start) {
        List<String> visitedorder = new ArrayList<>();
        Queue<String> queue = new LinkedList<>();
        Map<String, String> parentMap = new HashMap<>();
        queue.add(start);
        parentMap.put(start, null);
        while (!queue.isEmpty()) {
            String node = queue.poll();
            if (!visitedorder.contains(node)) {
                visitedorder.add(node);
                for (String Nrighbour : graph.get(node)) {
                    if (!parentMap.containsKey(Nrighbour)) {
                        parentMap.put(Nrighbour, node);
                        queue.add(Nrighbour);

                    }
                }
            }
        }
        return new BFSresult(visitedorder, parentMap);
    }

    public static List<String> getPath(Map<String, String> parentMap, String node) {
        List<String> path = new ArrayList<>();
        while (node != null) {
            path.add(node);
            node = parentMap.get(node);
        }
        Collections.reverse(path);
        return path;
    }

    public static void main(String[] args) {
        Map<String, List<String>> graph = new HashMap<>();
        graph.put("A", Arrays.asList("B", "C"));
        graph.put("B", Arrays.asList("A", "D", "E"));
        graph.put("C", Arrays.asList("A", "F"));
        graph.put("D", Arrays.asList("B"));
        graph.put("E", Arrays.asList("B", "F"));
        graph.put("F", Arrays.asList("C", "E"));
        String startNode = "A";
        BFSresult result = bfs(graph, startNode);
        System.out.println("Visited Order: " + result.visitedorder);
        System.out.println("Parent Relationships: " + result.parentMap);
        System.out.println("Path from A to E: " + getPath(result.parentMap, "E"));
    }
}
